packages/ui/src/hooks/useKeyBoardShortcut.tsx

```diff
@@ -0,0 +1,192 @@
+		} else {
+			// on other platforms, mod = ctrl
+			ctrlMatch = modifiers.mod === event.ctrlKey;
+		}
+	}
+
+	// check explicit modifier matches (these override mod if specified)
+	if (modifiers.ctrl !== undefined) {
+		ctrlMatch = modifiers.ctrl === event.ctrlKey;
+	}
+	if (modifiers.meta !== undefined) {
+		metaMatch = modifiers.meta === event.metaKey;
+	}
+
+	const shiftMatch = modifiers.shift === undefined || modifiers.shift === event.shiftKey;
+	const altMatch = modifiers.alt === undefined || modifiers.alt === event.altKey;
+
+	return ctrlMatch && metaMatch && shiftMatch && altMatch;
+};
+
+export function KeyboardShortcutProvider({ children }: PropsWithChildren) {
+	const shortcutsRef = React.useRef<Map<string, KeyboardShortcut[]>>(new Map());
+	const platform = usePlatform();
+
+	React.useEffect(() => {
+		const handleKeyDown = (event: KeyboardEvent) => {
+			const key = event.key.toLowerCase();
+			const shortcuts = shortcutsRef.current.get(key) || [];
+			event.code;
+
+			const sortedShortcuts = [...shortcuts].sort((a, b) => b.priority - a.priority);
+
+			for (const shortcut of sortedShortcuts) {
+				const matching = isMatchingEvent(event, shortcut, platform);
+
+				if (matching) {
+					event.preventDefault();
+					shortcut.handler(event);
+					return;
+				}
+			}
+		};
+
+		document.addEventListener("keydown", handleKeyDown);
+		return () => document.removeEventListener("keydown", handleKeyDown);
+	}, [platform]);
+
+	const register = (shortcut: KeyboardShortcut) => {
+		const key = shortcut.key.toLowerCase();
+		const shortcuts = shortcutsRef.current.get(key) || [];
+		shortcuts.push(shortcut);
+		shortcutsRef.current.set(key, shortcuts);
+
+		return () => {
+			const currentShortcuts = shortcutsRef.current.get(key) || [];
+			const filtered = currentShortcuts.filter((s) => s !== shortcut);
+			if (filtered.length === 0) {
+				shortcutsRef.current.delete(key);
+			} else {
+				shortcutsRef.current.set(key, filtered);
+			}
+		};
+	};
+
+	return (
+		<KeyboardShortcutContext.Provider value={{ register }}>
+			{children}
+		</KeyboardShortcutContext.Provider>
+	);
+}
+
+type Modifiers = "Mod" | "Ctrl" | "Shift" | "Alt" | "Meta";
+
+// prettier-ignore
+type Keys = "a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" | "j" | "k" | "l" | "m" | "n" | "o" | "p" | "q" | "r" | "s" | "t" | "u" | "v" | "w" | "x" | "y" | "z" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" | "0" | "F1" | "F2" | "F3" | "F4" | "F5" | "F6" | "F7" | "F8" | "F9" | "F10" | "F11" | "F12" | "ArrowUp" | "ArrowDown" | "ArrowLeft" | "ArrowRight" | "Enter" | "Escape" | "Backspace" | "Tab" | "Space" | "Delete" | "Home" | "End" | "PageUp" | "PageDown" | "CapsLock" | "NumLock" | "ScrollLock" | "Pause" | "PrintScreen" | "Insert";
+
+export type KeyCombination =
+	| `${Modifiers}+${Keys}`
+	| `${Modifiers}+${Modifiers}+${Keys}`
+	| `${Modifiers}+${Modifiers}+${Modifiers}+${Keys}`
+	| `${Modifiers}+${Modifiers}+${Modifiers}+${Modifiers}+${Keys}`;
```

Comment:
niceish type that allows for fully typed `Mod+k` like strings. doesn't really cover every key (only english-y ones), but that's fine bc it's us defining the shortcuts, and we (mostly) use English keyboards